include/Target/Btor/BtorToBtorIRTranslation.h

```diff
@@ -0,0 +1,291 @@
+//===----------------------------------------------------------------------===//
+//
+// This provides registration calls for Btor dialect to Btor IR translation.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef TARGET_BTOR_BTORTOBTORIRTRANSLATION_H
+#define TARGET_BTOR_BTORTOBTORIRTRANSLATION_H
+
+#include "Dialect/Btor/IR/BtorDialect.h"
+#include "Dialect/Btor/IR/BtorOps.h"
+
+#include "mlir/IR/OwningOpRef.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+
+#include <vector>
+#include <map>
+
+#include "btor2parser/btor2parser.h"
+
+namespace mlir {
+class MLIRContext;
+class ModuleOp;
+
+namespace btor {
+
+/// Deserializes the given Btor module and creates a MLIR ModuleOp
+/// in the given `context`. Makes use of btor2parser.
+
+class Deserialize {
+
+ public:
+///===----------------------------------------------------------------------===//
+/// Constructors and Destructors
+///===----------------------------------------------------------------------===//
+
+  Deserialize(MLIRContext *context, const std::string &s) : m_context(context), 
+    m_builder(OpBuilder(m_context)), m_unknownLoc(UnknownLoc::get(m_context)) {
+        m_modelFile = fopen(s.c_str(), "r");
+    }
+
+  ~Deserialize() {
+      if (m_model) {
+        btor2parser_delete(m_model);
+      }
+      if (m_modelFile) {
+        fclose(m_modelFile);
+      }
+  }
+
+///===----------------------------------------------------------------------===//
+/// Parse btor2 file
+///===----------------------------------------------------------------------===//
+  
+  bool parseModelIsSuccessful();
+
+///===----------------------------------------------------------------------===//
+/// Create MLIR module
+///===----------------------------------------------------------------------===//
+  
+  Value getFromCacheById(const int64_t id) {
+    assert(valueAtIdIsInCache(id));
+    return m_cache.at(id);
+  }
+
+  void setCacheWithId(const int64_t id, const Value &value) {
+    assert(!valueAtIdIsInCache(id));
+    m_cache[id] = value;
+    assert(valueAtIdIsInCache(id));
+  }
+
+  void setCacheWithId(const int64_t id, Operation * op) {
+    // We never have to use the result of a btor line with no 
+    // return values since btor2 doesn't allow it
+    if (hasReturnValue(getLineById(id))) {
```

Comment:
This check can be done by the caller instead. Will change into an assert. Lines such as `constraint` and `bad`, for example, need to be processed but need not be cached since their values are not used afterwards